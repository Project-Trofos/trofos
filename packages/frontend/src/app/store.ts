import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
