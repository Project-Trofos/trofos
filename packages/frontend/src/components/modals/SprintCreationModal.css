.new-sprint-btn {
  margin-left: 16px;
}
