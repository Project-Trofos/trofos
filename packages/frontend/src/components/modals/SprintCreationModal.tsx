import React, { useEffect, useState } from 'react';
import { Button, Form, Input, message, Modal, Select, DatePicker } from 'antd';
import { useParams } from 'react-router-dom';
import dayjs from 'dayjs';
import { useAddSprintMutation, useUpdateSprintMutation } from '../../api/socket/sprintHooks';
import type { Sprint } from '../../api/sprint';
import type { SprintFormFields, SprintUpdatePayload, AutoSprintTypes } from '../../helpers/SprintModal.types';
import './SprintCreationModal.css';
import { STEP_PROP, StepTarget } from '../tour/TourSteps';

const DURATION = [
  { id: 1, name: '1 Week' },
  { id: 2, name: '2 Weeks' },
  { id: 3, name: '3 Weeks' },
  { id: 4, name: '4 Weeks' },
  { id: 0, name: 'custom' },
];

function SprintCreationModal(props: SprintCreationModalPropsTypes): JSX.Element {
  const { isModalVisible, setIsModalVisible, sprint, setSprint, latestSprint } = props;
  const params = useParams();
  const [form] = Form.useForm();
  const { RangePicker } = DatePicker;
  const { Option } = Select;
  const { TextArea } = Input;

  const [addSprint] = useAddSprintMutation();
  const [updateSprint] = useUpdateSprintMutation();

  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (!sprint) {
      return form.setFieldsValue({
        name: latestSprint.name,
        duration: latestSprint.duration,
        dates: latestSprint.dates,
        startDate: latestSprint.dates ? latestSprint.dates[0] : undefined,
        goals: undefined,
      });
    }

    return form.setFieldsValue({
      ...sprint,
      startDate: sprint?.start_date ? dayjs(sprint.start_date) : undefined,
      dates: sprint?.start_date && sprint?.end_date ? [dayjs(sprint.start_date), dayjs(sprint.end_date)] : undefined,
    });
  }, [sprint, latestSprint]);

  const showModal = () => {
    setIsModalVisible(true);
  };

  const handleOk = () => {
    setIsModalVisible(false);
    setSprint(undefined);
  };

  const handleCancel = () => {
    setIsModalVisible(false);
    setSprint(undefined);
  };

  const handleProcessDate = (data: SprintFormFields) => {
    const { duration, startDate, dates } = data;
    const updatedData = { ...data };

    if ((duration !== 0 && !startDate) || (duration === 0 && !dates)) {
      updatedData.dates = null;
      delete updatedData.startDate;
      return updatedData;
    }

    if (duration !== 0) {
      updatedData.dates = [
        dayjs(startDate).toString(),
        dayjs(startDate)
          .add(duration * 7, 'day')
          .toString(),
      ];
      delete updatedData.startDate;
    }

    return updatedData;
  };

  const handleFormSubmit = async (data: SprintFormFields): Promise<void> => {
    setIsLoading(true);

    const processedPayload = handleProcessDate(data);

    const payload: SprintFormFields = {
      ...processedPayload,
      projectId: Number(params.projectId),
    };

    try {
      await addSprint(payload).unwrap();
      setIsModalVisible(false);
      form.resetFields();
      console.log('Success');
    } catch (e) {
      console.error(e);
    } finally {
      setIsLoading(false);
    }
  };

  const handleUpdateSprint = async (data: SprintFormFields): Promise<void> => {
    setIsLoading(true);

    if (!sprint) {
      return;
    }

    const processedPayload = handleProcessDate(data);

    const payload: SprintUpdatePayload = {
      ...processedPayload,
      sprintId: sprint.id,
      projectId: sprint.project_id,
    };

    try {
      await updateSprint(payload).unwrap();
      setIsModalVisible(false);
      setSprint(undefined);
      message.success('Sprint updated');
      console.log('Success');
    } catch (e) {
      console.error(e);
    } finally {
      setIsLoading(false);
    }
  };

  const renderFooter = (): JSX.Element[] => [
    <Button form="newSprint" key="submit" type="primary" htmlType="submit" loading={isLoading}>
      {sprint ? 'Update' : 'Create'}
    </Button>,
  ];

  const renderDatePicker = () => {
    if (form.getFieldValue('duration') === 0) {
      return (
        <Form.Item name="dates" label="Start and End Date">
          <RangePicker />
        </Form.Item>
      );
    }
    return (
      <Form.Item name="startDate" label="Start Date">
        <DatePicker />
      </Form.Item>
    );
  };

  const renderContent = (): JSX.Element => (
    <Form id="newSprint" form={form} onFinish={sprint ? handleUpdateSprint : handleFormSubmit} layout="vertical">
      <Form.Item name="name" rules={[{ required: true }]} label="Sprint Name">
        <Input />
      </Form.Item>
      <Form.Item name="duration" rules={[{ required: true }]} label="Duration">
        <Select>
          {DURATION.map((duration) => (
            <Option key={duration.id} value={duration.id}>
              {duration.name}
            </Option>
          ))}
        </Select>
      </Form.Item>
      <Form.Item noStyle shouldUpdate={(prevValues, currentValues) => prevValues.duration !== currentValues.duration}>
        {renderDatePicker}
      </Form.Item>
      <Form.Item name="goals" label="Sprint Goals">
        <TextArea autoSize={{ minRows: 5, maxRows: 8 }} />
      </Form.Item>
    </Form>
  );

  return (
    <>
      <Button
        className="new-sprint-btn"
        type="primary"
        onClick={showModal}
        {...{ [STEP_PROP]: StepTarget.NEW_SPRINT_BUTTON }}
      >
        New Sprint
      </Button>
      <Modal
        title="New Sprint"
        open={isModalVisible}
        onOk={handleOk}
        onCancel={handleCancel}
        footer={renderFooter()}
        width={600}
      >
        {renderContent()}
      </Modal>
    </>
  );
}

type SprintCreationModalPropsTypes = {
  isModalVisible: boolean;
  setIsModalVisible(isVisible: boolean): void;
  sprint?: Sprint;
  setSprint(sprint: Sprint | undefined): void;
  latestSprint: AutoSprintTypes;
};

SprintCreationModal.defaultProps = {
  sprint: undefined,
};

export default SprintCreationModal;
