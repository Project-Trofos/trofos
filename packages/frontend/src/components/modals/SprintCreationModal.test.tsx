import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import SprintCreationModal from './SprintCreationModal';
import store from '../../app/store';

describe('SprintModal tests', () => {
  it('renders new sprint modal with correct fields', () => {
    const mockSprintCreationModalProps = {
      isModalVisible: true,
      setIsModalVisible: vi.fn(),
      setSprint: vi.fn(),
      latestSprint: {
        name: 'Sprint 1',
        dates: undefined,
        duration: 1,
      },
    };

    const { baseElement } = render(
      <Provider store={store}>
        <SprintCreationModal
          isModalVisible={mockSprintCreationModalProps.isModalVisible}
          setIsModalVisible={mockSprintCreationModalProps.setIsModalVisible}
          setSprint={mockSprintCreationModalProps.setSprint}
          latestSprint={mockSprintCreationModalProps.latestSprint}
        />
      </Provider>,
    );

    // Ensure fields are present
    expect(screen.getByLabelText('Sprint Name')).toBeInTheDocument();
    expect(screen.getByLabelText('Duration')).toBeInTheDocument();
    expect(screen.getByLabelText('Start Date')).toBeInTheDocument();
    expect(screen.getByLabelText('Sprint Goals')).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });
});
