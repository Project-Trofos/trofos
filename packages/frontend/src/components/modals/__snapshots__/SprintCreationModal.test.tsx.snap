// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SprintModal tests > renders new sprint modal with correct fields 1`] = `
<body>
  <div>
    <button
      class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid new-sprint-btn"
      type="button"
    >
      <span>
        New Sprint
      </span>
    </button>
  </div>
  <div>
    <div
      class="ant-modal-root"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 600px;"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  New Sprint
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <form
                  class="ant-form ant-form-vertical"
                  id="newSprint"
                >
                  <div
                    class="ant-form-item ant-form-item-has-success"
                  >
                    <div
                      class="ant-row ant-form-item-row"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class="ant-form-item-required"
                          for="name"
                          title="Sprint Name"
                        >
                          Sprint Name
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <input
                              aria-required="true"
                              class="ant-input ant-input-outlined ant-input-status-success"
                              id="name"
                              type="text"
                              value="Sprint 1"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-form-item ant-form-item-has-success"
                  >
                    <div
                      class="ant-row ant-form-item-row"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class="ant-form-item-required"
                          for="duration"
                          title="Duration"
                        >
                          Duration
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <div
                              aria-required="true"
                              class="ant-select ant-select-outlined ant-select-in-form-item ant-select-status-success ant-select-single ant-select-show-arrow"
                            >
                              <div
                                class="ant-select-selector"
                              >
                                <span
                                  class="ant-select-selection-wrap"
                                >
                                  <span
                                    class="ant-select-selection-search"
                                  >
                                    <input
                                      aria-autocomplete="list"
                                      aria-controls="duration_list"
                                      aria-expanded="false"
                                      aria-haspopup="listbox"
                                      aria-owns="duration_list"
                                      aria-required="true"
                                      autocomplete="off"
                                      class="ant-select-selection-search-input"
                                      id="duration"
                                      readonly=""
                                      role="combobox"
                                      style="opacity: 0;"
                                      type="search"
                                      unselectable="on"
                                      value=""
                                    />
                                  </span>
                                  <span
                                    class="ant-select-selection-item"
                                    title="1 Week"
                                  >
                                    1 Week
                                  </span>
                                </span>
                              </div>
                              <span
                                aria-hidden="true"
                                class="ant-select-arrow"
                                style="user-select: none;"
                                unselectable="on"
                              >
                                <span
                                  aria-label="down"
                                  class="anticon anticon-down ant-select-suffix"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="64 64 896 896"
                                    width="1em"
                                  >
                                    <path
                                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-form-item"
                  >
                    <div
                      class="ant-row ant-form-item-row"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class=""
                          for="startDate"
                          title="Start Date"
                        >
                          Start Date
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <div
                              class="ant-picker ant-picker-outlined"
                            >
                              <div
                                class="ant-picker-input"
                              >
                                <input
                                  aria-invalid="false"
                                  autocomplete="off"
                                  id="startDate"
                                  placeholder="Select date"
                                  size="12"
                                  value=""
                                />
                                <span
                                  class="ant-picker-suffix"
                                >
                                  <span
                                    aria-label="calendar"
                                    class="anticon anticon-calendar"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="calendar"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-form-item"
                  >
                    <div
                      class="ant-row ant-form-item-row"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class=""
                          for="goals"
                          title="Sprint Goals"
                        >
                          Sprint Goals
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <textarea
                              class="ant-input ant-input-outlined"
                              id="goals"
                              style="resize: none; height: -54px; min-height: -30px; max-height: -54px;"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
              <div
                class="ant-modal-footer"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  form="newSprint"
                  type="submit"
                >
                  <span>
                    Create
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
  <textarea
    aria-hidden="true"
    name="hiddenTextarea"
    style="letter-spacing:normal;line-height:1.5714285714285714;padding-top:4px;padding-bottom:4px;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,'Helvetica Neue',Arial,'Noto Sans',sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol','Noto Color Emoji';font-weight:;font-size:14px;font-variant:;text-rendering:auto;text-transform:none;width:100%;text-indent:0;padding-left:11px;padding-right:11px;border-width:1px;box-sizing:border-box;word-break:;white-space:pre-wrap;
  min-height:0 !important;
  max-height:none !important;
  height:0 !important;
  visibility:hidden !important;
  overflow:hidden !important;
  position:absolute !important;
  z-index:-1000 !important;
  top:0 !important;
  right:0 !important;
  pointer-events: none !important;
"
    tab-index="-1"
  />
</body>
`;
